File: apps/web/lib/lms.ts
export type CourseSummary = {
  course_id: string;
  section_id: string;
  course_code: string;
  course_name: string;
  section_code: string;
  semester: string;
  lecturer_name?: string | null;
};

export type CourseDetail = {
  course_id: string;
  section_id: string;
  course_code: string;
  course_name: string;
  description: string;
  credits: number;
  section_code: string;
  semester: string;
  lecturer_name?: string | null;
};

export type Lesson = {
  id: string;
  section_id: string;
  title: string;
  content: string;
  order_index: number;
  is_published: boolean;
  created_at: string;
};

export type Assignment = {
  id: string;
  section_id: string;
  title: string;
  description: string;
  due_at?: string | null;
  max_score: number;
  created_at: string;
};

export type Submission = {
  id: string;
  assignment_id: string;
  student_id: string;
  student_name?: string | null;
  content: string;
  submitted_at: string;
  status: string;
  score?: number | null;
  feedback?: string | null;
};

export type Quiz = {
  id: string;
  section_id: string;
  title: string;
  duration_minutes: number;
  max_score: number;
  is_published: boolean;
  created_at: string;
};

export type QuizQuestion = {
  id: string;
  quiz_id: string;
  prompt: string;
  options: string[];
  points: number;
  order_index: number;
};

export type QuizAttemptResult = {
  attempt_id: string;
  quiz_id: string;
  score: number;
  max_score: number;
  submitted_at: string;
};

export type GradeItem = {
  id: string;
  source_type: string;
  source_id: string;
  source_title: string;
  score: number;
  feedback: string;
  section_id: string;
  student_id: string;
  student_name?: string | null;
  created_at: string;
};

export type AttendanceSession = {
  id: string;
  section_id: string;
  session_date: string;
  title: string;
  lecturer_id?: string | null;
};

export type AttendanceRecord = {
  id: string;
  section_id: string;
  student_id: string;
  student_name?: string | null;
  attendance_date: string;
  status: "present" | "absent" | "late" | string;
  note: string;
};

export type SectionStudent = {
  student_id: string;
  full_name: string;
  email: string;
};

export type MetaPage = {
  page: number;
  limit: number;
  total: number;
};

export type RoleItem = {
  id: string;
  code: "student" | "lecturer" | "admin" | "academic_staff" | "advisor";
  name: string;
  description: string;
};

export type PermissionItem = {
  id: string;
  code: string;
  name: string;
  module: string;
  description: string;
};

export type DepartmentItem = {
  id: string;
  code: string;
  name: string;
  description: string;
  created_at: string;
};

export type ProgramItem = {
  id: string;
  code: string;
  name: string;
  department_id?: string | null;
  total_credits: number;
  created_at: string;
};

export type CourseItem = {
  id: string;
  code: string;
  name: string;
  department_id?: string | null;
  credits: number;
  description: string;
  created_at: string;
};

export type SemesterItem = {
  id: string;
  code: string;
  name: string;
  start_date: string;
  end_date: string;
  status: string;
  created_at: string;
};

export type SectionItem = {
  id: string;
  course_id: string;
  lecturer_id?: string | null;
  code: string;
  semester: string;
  max_students: number;
  status: string;
  created_at: string;
};

export type CurriculumItem = {
  id: string;
  program_id: string;
  course_id: string;
  semester_no: number;
  is_required: boolean;
  created_at: string;
};

export type LecturerAssignmentItem = {
  id: string;
  section_id: string;
  lecturer_id: string;
  assigned_by?: string | null;
  role: string;
  created_at: string;
};

export type StudentTrackingItem = {
  student_id: string;
  student_name: string;
  student_email: string;
  section_id: string;
  section_code: string;
  average_score: number;
  attendance_present: number;
  attendance_total: number;
};

export type GradeApprovalItem = {
  id: string;
  section_id: string;
  requested_by?: string | null;
  approved_by?: string | null;
  status: string;
  note: string;
  created_at: string;
};

export type ExamApprovalItem = {
  id: string;
  quiz_id: string;
  requested_by?: string | null;
  approved_by?: string | null;
  status: string;
  note: string;
  created_at: string;
};

export type AdvisorStudentItem = {
  id: string;
  advisor_id: string;
  student_id: string;
  created_at: string;
};

export type StudentProgressItem = {
  student_id: string;
  student_name: string;
  student_email: string;
  total_grades: number;
  average_score: number;
  total_attendance: number;
  present_attendance: number;
  risk_alerts_open: number;
};

export type RiskAlertItem = {
  id: string;
  student_id: string;
  advisor_id?: string | null;
  risk_level: string;
  reason: string;
  status: string;
  recommended_action: string;
  created_at: string;
};

export type ConsultationItem = {
  id: string;
  advisor_id?: string | null;
  student_id: string;
  summary: string;
  action_plan: string;
  follow_up_at?: string | null;
  created_at: string;
};

export type StudyPlanItem = {
  id: string;
  student_id: string;
  advisor_id?: string | null;
  title: string;
  goals: string[];
  tasks: { [key: string]: string }[];
  created_at: string;
};

export type SupportRequestItem = {
  id: string;
  student_id: string;
  advisor_id?: string | null;
  title: string;
  description: string;
  status: string;
  escalation_note: string;
  created_at: string;
};

export type ForumTopicItem = {
  id: string;
  section_id: string;
  created_by?: string | null;
  title: string;
  content: string;
  status: string;
  is_pinned: boolean;
  replies_count: number;
  created_at: string;
};

export type LiveClassSessionItem = {
  id: string;
  section_id: string;
  lecturer_id?: string | null;
  title: string;
  scheduled_at: string;
  platform: string;
  meeting_url: string;
  recording_url: string;
  status: string;
  created_at: string;
};

export type EnrollmentActionResult = {
  requested: number;
  success: number;
  skipped: number;
  action_type: string;
  target_section_id: string;
  reason: string;
};

export type QualitySurveyItem = {
  id: string;
  title: string;
  category: string;
  department_id?: string | null;
  status: string;
  responses_count: number;
  created_at: string;
};

export type NotificationItem = {
  id: string;
  user_id: string;
  title: string;
  message: string;
  channel: string;
  is_read: boolean;
  created_at: string;
};

export type AnnouncementItem = {
  id: string;
  title: string;
  content: string;
  target_role: string;
  section_id?: string | null;
  department_id?: string | null;
  created_by?: string | null;
  created_at: string;
};

export type AuditLogItem = {
  id: string;
  actor_id?: string | null;
  action: string;
  resource_type: string;
  resource_id: string;
  detail: Record<string, string>;
  created_at: string;
};

export type UploadedFileItem = {
  id: string;
  owner_id?: string | null;
  module: string;
  original_name: string;
  stored_name: string;
  content_type: string;
  file_size: number;
  path: string;
  created_at: string;
};

export type RoleReport = {
  role: string;
  title: string;
  generated_at: string;
  items: { key: string; value: string }[];
};

export type AdminUserItem = {
  id: string;
  email: string;
  full_name: string;
  role: "student" | "lecturer" | "admin" | "academic_staff" | "advisor";
  department_id?: string | null;
  program_id?: string | null;
  is_active: boolean;
};

File: apps/api/app/models/lms.py
from __future__ import annotations

from datetime import datetime
from uuid import uuid4

from sqlalchemy import JSON, Boolean, DateTime, Float, ForeignKey, Integer, String, Text, UniqueConstraint
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base


def uuid_str() -> str:
    return str(uuid4())


class TimestampMixin:
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=datetime.utcnow)
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=datetime.utcnow,
        onupdate=datetime.utcnow,
    )


class RolePermission(Base):
    __tablename__ = "role_permissions"
    __table_args__ = (UniqueConstraint("role_id", "permission_id", name="uq_role_permission"),)

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=uuid_str)
    role_id: Mapped[str] = mapped_column(ForeignKey("roles.id", ondelete="CASCADE"), index=True)
    permission_id: Mapped[str] = mapped_column(ForeignKey("permissions.id", ondelete="CASCADE"), index=True)


class UserRoleLink(Base):
    __tablename__ = "user_roles"
    __table_args__ = (UniqueConstraint("user_id", "role_id", name="uq_user_role"),)

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=uuid_str)
    user_id: Mapped[str] = mapped_column(ForeignKey("users.id", ondelete="CASCADE"), index=True)
    role_id: Mapped[str] = mapped_column(ForeignKey("roles.id", ondelete="CASCADE"), index=True)


class Role(Base, TimestampMixin):
    __tablename__ = "roles"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=uuid_str)
    code: Mapped[str] = mapped_column(String(50), unique=True, nullable=False, index=True)
    name: Mapped[str] = mapped_column(String(120), nullable=False)
    description: Mapped[str] = mapped_column(String(255), default="")


class Permission(Base, TimestampMixin):
    __tablename__ = "permissions"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=uuid_str)
    code: Mapped[str] = mapped_column(String(120), unique=True, index=True)
    name: Mapped[str] = mapped_column(String(120))
    module: Mapped[str] = mapped_column(String(120))
    description: Mapped[str] = mapped_column(String(255), default="")


class Department(Base, TimestampMixin):
    __tablename__ = "departments"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=uuid_str)
    code: Mapped[str] = mapped_column(String(32), unique=True, index=True)
    name: Mapped[str] = mapped_column(String(120))
    description: Mapped[str] = mapped_column(String(255), default="")


class Program(Base, TimestampMixin):
    __tablename__ = "programs"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=uuid_str)
    department_id: Mapped[str] = mapped_column(ForeignKey("departments.id", ondelete="SET NULL"), nullable=True)
    code: Mapped[str] = mapped_column(String(32), unique=True, index=True)
    name: Mapped[str] = mapped_column(String(120))
    total_credits: Mapped[int] = mapped_column(Integer, default=120)


class Semester(Base, TimestampMixin):
    __tablename__ = "semesters"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=uuid_str)
    code: Mapped[str] = mapped_column(String(32), unique=True, index=True)
    name: Mapped[str] = mapped_column(String(120))
    start_date: Mapped[datetime] = mapped_column(DateTime(timezone=True))
    end_date: Mapped[datetime] = mapped_column(DateTime(timezone=True))
    status: Mapped[str] = mapped_column(String(32), default="planned")


class User(Base, TimestampMixin):
    __tablename__ = "users"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=uuid_str)
    email: Mapped[str] = mapped_column(String(150), unique=True, index=True)
    full_name: Mapped[str] = mapped_column(String(120))
    password_hash: Mapped[str] = mapped_column(String(255))
    role: Mapped[str] = mapped_column(String(50), index=True)
    department_id: Mapped[str] = mapped_column(ForeignKey("departments.id", ondelete="SET NULL"), nullable=True)
    program_id: Mapped[str] = mapped_column(ForeignKey("programs.id", ondelete="SET NULL"), nullable=True)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True)


class Course(Base, TimestampMixin):
    __tablename__ = "courses"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=uuid_str)
    code: Mapped[str] = mapped_column(String(32), unique=True, index=True)
    name: Mapped[str] = mapped_column(String(150))
    department_id: Mapped[str] = mapped_column(ForeignKey("departments.id", ondelete="SET NULL"), nullable=True)
    credits: Mapped[int] = mapped_column(Integer, default=3)
    description: Mapped[str] = mapped_column(Text, default="")


class CourseSection(Base, TimestampMixin):
    __tablename__ = "course_sections"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=uuid_str)
    course_id: Mapped[str] = mapped_column(ForeignKey("courses.id", ondelete="CASCADE"), index=True)
    lecturer_id: Mapped[str] = mapped_column(ForeignKey("users.id", ondelete="SET NULL"), nullable=True)
    code: Mapped[str] = mapped_column(String(32), unique=True, index=True)
    semester: Mapped[str] = mapped_column(String(32), default="2026A")
    max_students: Mapped[int] = mapped_column(Integer, default=60)
    status: Mapped[str] = mapped_column(String(32), default="draft")


class Enrollment(Base, TimestampMixin):
    __tablename__ = "enrollments"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=uuid_str)
    student_id: Mapped[str] = mapped_column(ForeignKey("users.id", ondelete="CASCADE"), index=True)
    section_id: Mapped[str] = mapped_column(ForeignKey("course_sections.id", ondelete="CASCADE"), index=True)
    enrollment_status: Mapped[str] = mapped_column(String(32), default="active")


class Lesson(Base, TimestampMixin):
    __tablename__ = "lessons"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=uuid_str)
    section_id: Mapped[str] = mapped_column(ForeignKey("course_sections.id", ondelete="CASCADE"), index=True)
    title: Mapped[str] = mapped_column(String(150))
    content: Mapped[str] = mapped_column(Text, default="")
    order_index: Mapped[int] = mapped_column(Integer, default=1)
    is_published: Mapped[bool] = mapped_column(Boolean, default=False)


class Assignment(Base, TimestampMixin):
    __tablename__ = "assignments"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=uuid_str)
    section_id: Mapped[str] = mapped_column(ForeignKey("course_sections.id", ondelete="CASCADE"), index=True)
    title: Mapped[str] = mapped_column(String(150))
    description: Mapped[str] = mapped_column(Text, default="")
    due_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=True)
    max_score: Mapped[float] = mapped_column(Float, default=10.0)


class Submission(Base, TimestampMixin):
    __tablename__ = "submissions"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=uuid_str)
    assignment_id: Mapped[str] = mapped_column(ForeignKey("assignments.id", ondelete="CASCADE"), index=True)
    student_id: Mapped[str] = mapped_column(ForeignKey("users.id", ondelete="CASCADE"), index=True)
    content: Mapped[str] = mapped_column(Text, default="")
    submitted_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=datetime.utcnow)
    status: Mapped[str] = mapped_column(String(32), default="submitted")


class Quiz(Base, TimestampMixin):
    __tablename__ = "quizzes"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=uuid_str)
    section_id: Mapped[str] = mapped_column(ForeignKey("course_sections.id", ondelete="CASCADE"), index=True)
    title: Mapped[str] = mapped_column(String(150))
    duration_minutes: Mapped[int] = mapped_column(Integer, default=30)
    max_score: Mapped[float] = mapped_column(Float, default=10.0)
    is_published: Mapped[bool] = mapped_column(Boolean, default=False)


class QuizQuestion(Base, TimestampMixin):
    __tablename__ = "quiz_questions"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=uuid_str)
    quiz_id: Mapped[str] = mapped_column(ForeignKey("quizzes.id", ondelete="CASCADE"), index=True)
    prompt: Mapped[str] = mapped_column(Text)
    question_type: Mapped[str] = mapped_column(String(32), default="multiple_choice")
    options: Mapped[list[str]] = mapped_column(JSON, default=list)
    correct_answer: Mapped[str] = mapped_column(String(255), default="")
    points: Mapped[float] = mapped_column(Float, default=1.0)
    order_index: Mapped[int] = mapped_column(Integer, default=1)


class QuizAttempt(Base, TimestampMixin):
    __tablename__ = "quiz_attempts"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=uuid_str)
    quiz_id: Mapped[str] = mapped_column(ForeignKey("quizzes.id", ondelete="CASCADE"), index=True)
    student_id: Mapped[str] = mapped_column(ForeignKey("users.id", ondelete="CASCADE"), index=True)
    answers: Mapped[dict[str, str]] = mapped_column(JSON, default=dict)
    score: Mapped[float] = mapped_column(Float, default=0)
    submitted_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=datetime.utcnow)


class Grade(Base, TimestampMixin):
    __tablename__ = "grades"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=uuid_str)
    student_id: Mapped[str] = mapped_column(ForeignKey("users.id", ondelete="CASCADE"), index=True)
    section_id: Mapped[str] = mapped_column(ForeignKey("course_sections.id", ondelete="CASCADE"), index=True)
    source_type: Mapped[str] = mapped_column(String(32), default="assignment")
    source_id: Mapped[str] = mapped_column(String(36), index=True)
    score: Mapped[float] = mapped_column(Float, default=0)
    feedback: Mapped[str] = mapped_column(Text, default="")


class AttendanceRecord(Base, TimestampMixin):
    __tablename__ = "attendance_records"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=uuid_str)
    student_id: Mapped[str] = mapped_column(ForeignKey("users.id", ondelete="CASCADE"), index=True)
    section_id: Mapped[str] = mapped_column(ForeignKey("course_sections.id", ondelete="CASCADE"), index=True)
    attendance_date: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=datetime.utcnow)
    status: Mapped[str] = mapped_column(String(32), default="present")
    note: Mapped[str] = mapped_column(String(255), default="")


class AttendanceSession(Base, TimestampMixin):
    __tablename__ = "attendance_sessions"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=uuid_str)
    section_id: Mapped[str] = mapped_column(ForeignKey("course_sections.id", ondelete="CASCADE"), index=True)
    lecturer_id: Mapped[str] = mapped_column(ForeignKey("users.id", ondelete="SET NULL"), nullable=True)
    session_date: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=datetime.utcnow)
    title: Mapped[str] = mapped_column(String(160), default="Attendance Session")


class Notification(Base, TimestampMixin):
    __tablename__ = "notifications"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=uuid_str)
    user_id: Mapped[str] = mapped_column(ForeignKey("users.id", ondelete="CASCADE"), index=True)
    title: Mapped[str] = mapped_column(String(160))
    message: Mapped[str] = mapped_column(Text)
    channel: Mapped[str] = mapped_column(String(32), default="in_app")
    is_read: Mapped[bool] = mapped_column(Boolean, default=False)


class Announcement(Base, TimestampMixin):
    __tablename__ = "announcements"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=uuid_str)
    title: Mapped[str] = mapped_column(String(180))
    content: Mapped[str] = mapped_column(Text)
    target_role: Mapped[str] = mapped_column(String(50), default="all")
    section_id: Mapped[str] = mapped_column(ForeignKey("course_sections.id", ondelete="SET NULL"), nullable=True)
    department_id: Mapped[str] = mapped_column(ForeignKey("departments.id", ondelete="SET NULL"), nullable=True)
    created_by: Mapped[str] = mapped_column(ForeignKey("users.id", ondelete="SET NULL"), nullable=True)


class CurriculumEntry(Base, TimestampMixin):
    __tablename__ = "curriculum_entries"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=uuid_str)
    program_id: Mapped[str] = mapped_column(ForeignKey("programs.id", ondelete="CASCADE"), index=True)
    course_id: Mapped[str] = mapped_column(ForeignKey("courses.id", ondelete="CASCADE"), index=True)
    semester_no: Mapped[int] = mapped_column(Integer, default=1)
    is_required: Mapped[bool] = mapped_column(Boolean, default=True)


class LecturerAssignment(Base, TimestampMixin):
    __tablename__ = "lecturer_assignments"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=uuid_str)
    section_id: Mapped[str] = mapped_column(ForeignKey("course_sections.id", ondelete="CASCADE"), index=True)
    lecturer_id: Mapped[str] = mapped_column(ForeignKey("users.id", ondelete="CASCADE"), index=True)
    assigned_by: Mapped[str] = mapped_column(ForeignKey("users.id", ondelete="SET NULL"), nullable=True)
    role: Mapped[str] = mapped_column(String(50), default="primary_lecturer")


class GradeApproval(Base, TimestampMixin):
    __tablename__ = "grade_approvals"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=uuid_str)
    section_id: Mapped[str] = mapped_column(ForeignKey("course_sections.id", ondelete="CASCADE"), index=True)
    requested_by: Mapped[str] = mapped_column(ForeignKey("users.id", ondelete="SET NULL"), nullable=True)
    approved_by: Mapped[str] = mapped_column(ForeignKey("users.id", ondelete="SET NULL"), nullable=True)
    status: Mapped[str] = mapped_column(String(32), default="pending")
    note: Mapped[str] = mapped_column(Text, default="")


class ExamApproval(Base, TimestampMixin):
    __tablename__ = "exam_approvals"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=uuid_str)
    quiz_id: Mapped[str] = mapped_column(ForeignKey("quizzes.id", ondelete="CASCADE"), index=True)
    requested_by: Mapped[str] = mapped_column(ForeignKey("users.id", ondelete="SET NULL"), nullable=True)
    approved_by: Mapped[str] = mapped_column(ForeignKey("users.id", ondelete="SET NULL"), nullable=True)
    status: Mapped[str] = mapped_column(String(32), default="pending")
    note: Mapped[str] = mapped_column(Text, default="")


class AdvisorStudent(Base, TimestampMixin):
    __tablename__ = "advisor_students"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=uuid_str)
    advisor_id: Mapped[str] = mapped_column(ForeignKey("users.id", ondelete="CASCADE"), index=True)
    student_id: Mapped[str] = mapped_column(ForeignKey("users.id", ondelete="CASCADE"), index=True)
    __table_args__ = (UniqueConstraint("advisor_id", "student_id", name="uq_advisor_student"),)


class RiskAlert(Base, TimestampMixin):
    __tablename__ = "risk_alerts"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=uuid_str)
    student_id: Mapped[str] = mapped_column(ForeignKey("users.id", ondelete="CASCADE"), index=True)
    advisor_id: Mapped[str] = mapped_column(ForeignKey("users.id", ondelete="SET NULL"), nullable=True)
    risk_level: Mapped[str] = mapped_column(String(32), default="medium")
    reason: Mapped[str] = mapped_column(Text)
    status: Mapped[str] = mapped_column(String(32), default="open")
    recommended_action: Mapped[str] = mapped_column(Text, default="")


class ConsultationRecord(Base, TimestampMixin):
    __tablename__ = "consultation_records"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=uuid_str)
    advisor_id: Mapped[str] = mapped_column(ForeignKey("users.id", ondelete="SET NULL"), nullable=True, index=True)
    student_id: Mapped[str] = mapped_column(ForeignKey("users.id", ondelete="CASCADE"), index=True)
    summary: Mapped[str] = mapped_column(Text, default="")
    action_plan: Mapped[str] = mapped_column(Text, default="")
    follow_up_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=True)


class StudyPlan(Base, TimestampMixin):
    __tablename__ = "study_plans"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=uuid_str)
    student_id: Mapped[str] = mapped_column(ForeignKey("users.id", ondelete="CASCADE"), index=True)
    advisor_id: Mapped[str] = mapped_column(ForeignKey("users.id", ondelete="SET NULL"), nullable=True, index=True)
    title: Mapped[str] = mapped_column(String(160))
    goals: Mapped[list[str]] = mapped_column(JSON, default=list)
    tasks: Mapped[list[dict[str, str]]] = mapped_column(JSON, default=list)


class SupportRequest(Base, TimestampMixin):
    __tablename__ = "support_requests"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=uuid_str)
    student_id: Mapped[str] = mapped_column(ForeignKey("users.id", ondelete="CASCADE"), index=True)
    advisor_id: Mapped[str] = mapped_column(ForeignKey("users.id", ondelete="SET NULL"), nullable=True, index=True)
    title: Mapped[str] = mapped_column(String(160))
    description: Mapped[str] = mapped_column(Text, default="")
    status: Mapped[str] = mapped_column(String(32), default="open")
    escalation_note: Mapped[str] = mapped_column(Text, default="")


class ForumTopic(Base, TimestampMixin):
    __tablename__ = "forum_topics"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=uuid_str)
    section_id: Mapped[str] = mapped_column(ForeignKey("course_sections.id", ondelete="CASCADE"), index=True)
    created_by: Mapped[str] = mapped_column(ForeignKey("users.id", ondelete="SET NULL"), nullable=True, index=True)
    title: Mapped[str] = mapped_column(String(180))
    content: Mapped[str] = mapped_column(Text, default="")
    status: Mapped[str] = mapped_column(String(32), default="open")
    is_pinned: Mapped[bool] = mapped_column(Boolean, default=False)
    replies_count: Mapped[int] = mapped_column(Integer, default=0)


class LiveClassSession(Base, TimestampMixin):
    __tablename__ = "live_class_sessions"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=uuid_str)
    section_id: Mapped[str] = mapped_column(ForeignKey("course_sections.id", ondelete="CASCADE"), index=True)
    lecturer_id: Mapped[str] = mapped_column(ForeignKey("users.id", ondelete="SET NULL"), nullable=True, index=True)
    title: Mapped[str] = mapped_column(String(180))
    scheduled_at: Mapped[datetime] = mapped_column(DateTime(timezone=True))
    platform: Mapped[str] = mapped_column(String(64), default="zoom")
    meeting_url: Mapped[str] = mapped_column(String(255), default="")
    recording_url: Mapped[str] = mapped_column(String(255), default="")
    status: Mapped[str] = mapped_column(String(32), default="scheduled")


class QualitySurvey(Base, TimestampMixin):
    __tablename__ = "quality_surveys"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=uuid_str)
    title: Mapped[str] = mapped_column(String(180))
    category: Mapped[str] = mapped_column(String(64), default="course")
    department_id: Mapped[str] = mapped_column(ForeignKey("departments.id", ondelete="SET NULL"), nullable=True, index=True)
    status: Mapped[str] = mapped_column(String(32), default="open")
    responses_count: Mapped[int] = mapped_column(Integer, default=0)


class AuditLog(Base):
    __tablename__ = "audit_logs"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=uuid_str)
    actor_id: Mapped[str] = mapped_column(ForeignKey("users.id", ondelete="SET NULL"), nullable=True, index=True)
    action: Mapped[str] = mapped_column(String(120), index=True)
    resource_type: Mapped[str] = mapped_column(String(120))
    resource_id: Mapped[str] = mapped_column(String(36), default="")
    detail: Mapped[dict[str, str]] = mapped_column(JSON, default=dict)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=datetime.utcnow)


class UploadedFile(Base, TimestampMixin):
    __tablename__ = "uploaded_files"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=uuid_str)
    owner_id: Mapped[str] = mapped_column(ForeignKey("users.id", ondelete="SET NULL"), nullable=True, index=True)
    module: Mapped[str] = mapped_column(String(120), default="general")
    original_name: Mapped[str] = mapped_column(String(255))
    stored_name: Mapped[str] = mapped_column(String(255))
    content_type: Mapped[str] = mapped_column(String(120), default="application/octet-stream")
    file_size: Mapped[int] = mapped_column(Integer, default=0)
    path: Mapped[str] = mapped_column(String(255))
